### PR TITLE
Feature/380 merge extra java options

### LINKED ISSUE #380: Merge extraJavaOptions from template, job and properties file
**Current situation**
With #366 extraJavaOptions from the job definition and options globally defined in the properties file are merged. However, it is also possible to define extraJavaOptions in job templates. Currently, extraJavaOptions defined in a job template are dropped if extraJavaOptions are defined in the job definition.
image

image

image

**Desired behaviour**
If a job template defines additional spark config keys `spark.executor.extraJavaOptions` or `spark.driver.extraJavaOptions`, the value from either property should not be overwritten by the job definition in any case, but merged with the `spark.executor.extraJavaOptions` and `spark.driver.extraJavaOptions` of the job definition.

The values should be concatenated. The job definition's extraJavaOptions should be appended to the job template's extraJavaOptions such that anything defined in the job definition's extraJavaOptions has precedence over the job template.

### PR DESCRIPTION
**Pull request by kevinwallimann**
Closes #380

### CODE REVIEW

File: src/main/scala/za/co/absa/hyperdrive/trigger/scheduler/utilities/Configs.scala [hunk 1]
```diff
@@ -100,4 +100,9 @@ object SparkExecutorConfig {
     Try(Configs.conf.getString("sparkYarnSink.filesToDeploy").split(",").toSeq).getOrElse(Seq.empty[String])
   def getAdditionalConfs: Map[String, String] =
     Configs.getMapFromConf("sparkYarnSink.additionalConfs")
+}
+
+object JobDefinitionConfig {
+  val KeysToMerge = Set("spark.executor.extraJavaOptions", "spark.driver.extraJavaOptions")
```
Comment (jozefbakus) [inline src/main/scala/za/co/absa/hyperdrive/trigger/scheduler/utilities/Configs.scala:106]: I would move this to SparkExecutorConfig

File: src/main/scala/za/co/absa/hyperdrive/trigger/scheduler/executors/spark/SparkExecutor.scala [hunk 2]
```diff
@@ -92,17 +92,17 @@ object SparkExecutor extends Executor {
     sparkParameters.additionalJars.foreach(additionalJar => sparkLauncher.addJar(additionalJar))
     sparkParameters.additionalFiles.foreach(additionalFile => sparkLauncher.addFile(additionalFile))
     sparkParameters.additionalSparkConfig.foreach(conf => sparkLauncher.setConf(conf._1, conf._2))
-    mergeAdditionalSparkConfig(SparkExecutorConfig.getAdditionalConfs, sparkParameters.additionalSparkConfig, keysToMerge)
+    mergeAdditionalSparkConfig(SparkExecutorConfig.getAdditionalConfs, sparkParameters.additionalSparkConfig)
```
Comment (jozefbakus) [inline src/main/scala/za/co/absa/hyperdrive/trigger/scheduler/executors/spark/SparkExecutor.scala:95]: Does this override configs from lines above ? 
    SparkExecutorConfig.getAdditionalConfs.foreach(conf => sparkLauncher.setConf(conf._1, conf._2))
and
    sparkParameters.additionalSparkConfig.foreach(conf => sparkLauncher.setConf(conf._1, conf._2))

File: src/test/scala/za/co/absa/hyperdrive/trigger/api/rest/utils/JobTemplateResolutionUtilTest.scala [hunk 3]
```diff
@@ -170,6 +170,35 @@ class JobTemplateResolutionUtilTest extends FlatSpec with Matchers {
     )
```
Comment (jozefbakus) [inline src/test/scala/za/co/absa/hyperdrive/trigger/api/rest/utils/JobTemplateResolutionUtilTest.scala:170]: Hmm is this behavior correct ? It is the same for (JobTemplate and SparkExecutor)
additionalSparkConfig: "spark.driver.extraJavaOptions" -> "-user.prop=userDriver -aaa.bbb=xxx"
templateParameters: "spark.driver.extraJavaOptions" -> "-template.prop=templateDriver -aaa.bbb=yyy"
Result is:
"spark.driver.extraJavaOptions" -> "-template.prop=templateDriver -aaa.bbb=yyy -user.prop=userDriver -aaa.bbb=xxx"

Comment (kevinwallimann) [reply src/test/scala/za/co/absa/hyperdrive/trigger/api/rest/utils/JobTemplateResolutionUtilTest.scala:170]: Yes, it's the desired behavior. Both the template properties and the user properties should be included in the final extraJavaOptions. If you want to override a template option, the assumption is that for java options, the succeeding options override the preceding options, that's why the user options are after the template options

File: src/main/scala/za/co/absa/hyperdrive/trigger/scheduler/executors/spark/SparkExecutor.scala (hunk 2 above)
Comment (kevinwallimann) [reply src/main/scala/za/co/absa/hyperdrive/trigger/scheduler/executors/spark/SparkExecutor.scala:95]: Yes, but it only overrides the `KeysToMerge` configs. I could also filter out the `KeysToMerge` in `sparkParameters.additionalSparkConfig.foreach`, but that would mainly just be for better readability, the end result wouldn't change.

File: src/main/scala/za/co/absa/hyperdrive/trigger/scheduler/utilities/Configs.scala (hunk 1 above)
Comment (kevinwallimann) [reply src/main/scala/za/co/absa/hyperdrive/trigger/scheduler/utilities/Configs.scala:106]: Yes, but it's not only about the SparkExecutor. It's also about the JobTemplate and JobDefinition. But anyway I can move it

Comment (jozefbakus) [reply src/main/scala/za/co/absa/hyperdrive/trigger/scheduler/utilities/Configs.scala:106]: Yes you are right. Leave it here

File: src/main/scala/za/co/absa/hyperdrive/trigger/scheduler/executors/spark/SparkExecutor.scala (hunk 2 above)
Comment (jozefbakus) [reply src/main/scala/za/co/absa/hyperdrive/trigger/scheduler/executors/spark/SparkExecutor.scala:95]: Yes, I thought so. You can leave it as it is.

**Review by jozefbakus — APPROVED** [review]

**Pull request merged**